services/src/tasks/mod.rs

```diff
@@ -0,0 +1,141 @@
+#[async_trait::async_trait]
+pub trait TaskDb<C: TaskContext>: Send + Sync {
+    async fn register(&self, task: Box<dyn Task<C>>) -> Result<TaskId, TaskError>;
+
+    async fn status(&self, task_id: TaskId) -> Result<TaskStatus, TaskError>;
+
+    async fn list(&self, options: TaskListOptions) -> Result<Vec<TaskStatusWithId>, TaskError>;
+}
+
+identifier!(TaskId);
+
+/// A task that can run asynchronously and reports its status.
+#[async_trait::async_trait]
+pub trait Task<C: TaskContext>: Send + Sync {
+    async fn run(self: Box<Self>, ctx: C) -> Result<Box<dyn TaskStatusInfo>, Box<dyn ErrorSource>>;
+
+    fn boxed(self) -> Box<dyn Task<C>>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+/// A way to supply status updates from a `Task` to a `TaskDb`.
+#[async_trait::async_trait]
+pub trait TaskContext: Send + Sync {
+    /// Set the completion percentage (%) of the task.
+    /// This is a number between 0 and 100.
+    ///
+    /// Moreover, set a status message.
+    ///
+    async fn set_completion(&self, pct_complete: u8, status: Box<dyn TaskStatusInfo>);
+}
+
+/// One of the statuses a `Task` can be in.
+#[derive(Debug, Clone, Serialize)]
+#[serde(rename_all = "camelCase", tag = "status")]
+pub enum TaskStatus {
+    // Pending, // TODO: at some point, don't just run things
+    Running(Arc<RunningTaskStatusInfo>),
+    Completed {
+        info: Arc<Box<dyn TaskStatusInfo>>,
+    },
+    #[serde(serialize_with = "serialize_failed_task_status")]
+    Failed {
+        error: Arc<Box<dyn ErrorSource>>,
+    },
+}
+
+#[derive(Debug, Clone, Serialize)]
+pub struct TaskStatusWithId {
+    pub task_id: TaskId,
+    #[serde(flatten)]
+    pub status: TaskStatus,
+}
+
+impl TaskStatus {
+    pub fn completed(info: Arc<Box<dyn TaskStatusInfo>>) -> Self {
+        TaskStatus::Completed { info }
+    }
+
+    pub fn failed(error: Arc<Box<dyn ErrorSource>>) -> Self {
+        TaskStatus::Failed { error }
+    }
+
+    pub fn is_running(&self) -> bool {
+        matches!(self, TaskStatus::Running(_))
+    }
+}
+
+#[derive(Debug, Serialize)]
+pub struct RunningTaskStatusInfo {
+    pct_complete: u8,
+    info: Box<dyn TaskStatusInfo>,
+}
+
+impl RunningTaskStatusInfo {
+    pub fn new(pct_complete: u8, info: Box<dyn TaskStatusInfo>) -> Arc<Self> {
+        Arc::new(RunningTaskStatusInfo { pct_complete, info })
+    }
+}
+
+#[allow(clippy::borrowed_box)]
+fn serialize_failed_task_status<S>(
+    error: &Box<dyn ErrorSource>,
+    serializer: S,
+) -> Result<S::Ok, S::Error>
+where
+    S: Serializer,
+{
+    serializer.serialize_str(error.to_string().as_str())
+}
+
+/// Trait for information about the status of a task.
+pub trait TaskStatusInfo: erased_serde::Serialize + Send + Sync + fmt::Debug {
+    fn boxed(self) -> Box<dyn TaskStatusInfo>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+erased_serde::serialize_trait_object!(TaskStatusInfo);
```

Comment:
why can you not use typetag?

Reply:
I can, but this leads to super weird infos, because every implementation, for instance, just a plain string needs a type definition. Here, it just needs to be serializable.

Reply:
the crate says

> This library should be considered a low-level building block for interacting with Serde APIs in an object-safe way. Most use cases will require higher level functionality such as provided by [typetag](https://github.com/dtolnay/typetag) which uses this crate internally.

so I am just wondering whether we want to depend on such a create. How much more work is it to use typetag instead? I have to admit I don't get how it works right now with erased_serde, but I haven't looked really into it either

Reply:
We already do since `typetag` is implemented using `erased_serde`. This is why I chose it, to not introduce something really new.